ConvolutionTranspose

vdumoulin opened the pull request:
This PR does the following:
* Implement `ConvolutionTranspose`, also colloquially called "deconvolution", like in [Zeiler et al., 2010](http://www.matthewzeiler.com/pubs/cvpr2010/cvpr2010.pdf). Currently implemented in Tensorflow under the name `conv2d_transpose`, which they describe as "the transpose (gradient) of `conv2d`".
* Implement `ConvolutionTransposeActivation`, in analogy to `ConvolutionActivation`.
* Switch to Theano's freshly-[renamed](https://github.com/Theano/Theano/pull/3679) `pool_2d` and `Pool`.
* Switch to Theano's recent `abstract_conv` interface. The convolution ops created via this interface are placeholders that get replaced at graph compilation time with the best option available.

I'm very open to feedback regarding design choices, names, etc. In particular, I'm not sure `Convolutional.conv2d_impl` and `Convolutional.get_output_shape` are needed anymore, but I'm afraid of breaking external code by removing these attributes.

dmitriy-serdyuk commented:
It looks, like this piece of code copy-pastes a lot from `Convolutional`. Is it possible to inherit from `Convolutional` and redefine the `conv2d_impl` to use the gradient?

According to the name convention: `ConvolutionTranspose`  is not consistent with `Convolutional`. I propose to use `Deconvolutional` and explain in the documentation what it does.

dwf commented:
"Deconvolution" is a huge misnomer though.

https://en.wikipedia.org/wiki/Deconvolution

vdumoulin commented:
I agree with @dwf there. The operation implemented in `ConvolutionTranspose` really *is* the transpose of the matrix representation of some convolution.

vdumoulin commented:
Well this is awkward. It looks like Theano doesn't like our compile flags when using `abstract_conv`.

dwf commented:
Perhaps we should add a test once that's resolved...

On Wed, Jan 20, 2016, 1:08 PM vdumoulin <notifications@github.com> wrote:

> Well this is awkward. It looks like Theano doesn't like our compile flags
> when using abstract_conv.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/946#issuecomment-173309062>.
>

vdumoulin commented:
Our tests are run in `FAST_COMPILE` mode to save time, but that forces the Python implementation to be used. The problem was that there is no Python implementation of `CorrMM`, which is used on Travis since `cudnn` isn't available.

The solution is to manually force the `FAST_RUN` mode to be used in the `theano.function` call for `blocks.bricks.conv` tests.

dwf commented:
Okay, can you leave a comment to that effect (even just copy and paste what
you just wrote) if you haven't already?

On Wed, Jan 20, 2016, 1:55 PM vdumoulin <notifications@github.com> wrote:

> Our tests are run in FAST_COMPILE mode to save time, but that forces the
> Python implementation to be used. The problem was that there is no Python
> implementation of CorrMM, which is used on Travis since cudnn isn't
> available.
>
> The solution is to manually force the FAST_RUN mode to be used in the
> theano.function call for blocks.bricks.conv tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/946#issuecomment-173323945>.
>

vdumoulin commented:
@dwf f588e4b

vdumoulin commented:
After further investigation, the `CorrMM` implementation should not have been used by Theano, since we disable C code through the `cxx=` flag.

This should be fixed when Theano/Theano#3905 is merged, at which point I'll re-run the tests to confirm that the issue is gone.

vdumoulin commented:
The CI tests are now passing!

dwf commented on blocks/bricks/conv.py:
```diff
@@ -177,6 +180,144 @@ def num_output_channels(self):
+        self.batch_size = batch_size
+        self.num_channels = num_channels
+        self.image_size = image_size
+        self.original_image_size = original_image_size
```
Is this strictly required? Seems kind of odd. Also looks like it's not required by `AbstractConv2d_gradInputs`.

vdumoulin replied:
Judging from the conversation I had with @lamblin regarding Theano/Theano#3913, it's not required, but there are some cases where the original size can be ambiguous with non-unit strides.

Take for example a `3x3` convolution with strides `2x2`. Both a `9x9` input and a `10x10` input will produce a `4x4` output, so the gradient of the output with respect to the input is undefined if we don't know the original input size.

dwf replied:
Is it simple to enumerate the cases where it's ambiguous? That, along with an optional argument, would be ideal. If not, a short explanation of why this is required would be good.

vdumoulin replied:
Pinging @lamblin on that.

lamblin replied:
So, maybe I'm too late, but the original size is ambiguous iff the stride is greater than 1.
However, currently the Theano Ops for `grad_wrt_inputs` require it as an input all the time, even when the stride is 1 and it could be inferred from the shapes of "topgrad" and "weights". Ideally, this would be improved in Theano.
Note that this computation when the stride is 1 would be similar to the shape inference for the forward convolution, but not exactly the same, since the padding would have to be handled differently.

dwf commented on blocks/bricks/conv.py:
```diff
@@ -177,6 +180,144 @@ def num_output_channels(self):
+        Theano convolution operator, possibly resulting in faster
+        execution.
+    image_size : tuple, optional
+        The height and width of the input (image or feature map). If given,
```
I think this description is wrong. It never gets passed, but it's necessary if you want untied biases.

vdumoulin replied:
I'll update it. Would you also like me to update the `Convolutional` description as well?

dwf replied:
Isn't it passed for Convolutional?

On Fri, Jan 22, 2016 at 5:40 PM, vdumoulin <notifications@github.com> wrote:

> In blocks/bricks/conv.py
> <https://github.com/mila-udem/blocks/pull/946#discussion_r50600970>:
>
> > +    num_filters : int
> > +        Number of filters per channel.
> > +    num_channels : int
> > +        Number of input channels in the image. For the first layer this is
> > +        normally 1 for grayscale images and 3 for color (RGB) images. For
> > +        subsequent layers this is equal to the number of filters output by
> > +        the previous convolutional layer. The filters are pooled over the
> > +        channels.
> > +    original_image_size : tuple
> > +        The height and width of the output (image or feature map).
> > +    batch_size : int, optional
> > +        Number of examples per batch. If given, this will be passed to
> > +        Theano convolution operator, possibly resulting in faster
> > +        execution.
> > +    image_size : tuple, optional
> > +        The height and width of the input (image or feature map). If given,
>
> I'll update it. Would you also like me to update the Convolutional
> description as well?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/946/files#r50600970>.
>

vdumoulin replied:
You're right.

vdumoulin replied:
b3450f4 addresses that as well as @dmitriy-serdyuk 's suggestion of making `ConvolutionalTranspose` a subclass of `Convolutional`.

dwf commented on blocks/bricks/conv.py:
```diff
@@ -177,6 +180,144 @@ def num_output_channels(self):
+        the previous convolutional layer. The filters are pooled over the
+        channels.
+    original_image_size : tuple
+        The height and width of the output (image or feature map).
```
This could be clearer. Maybe "The height and width of the image that forms the output of the transpose operation, which is the input of the original (non-transposed) convolution."

dwf commented on blocks/bricks/conv.py:
```diff
@@ -177,6 +180,144 @@ def num_output_channels(self):
+        normally 1 for grayscale images and 3 for color (RGB) images. For
+        subsequent layers this is equal to the number of filters output by
+        the previous convolutional layer. The filters are pooled over the
+        channels.
```
I'm confused as to whether this refers to the output (of the transpose operation) or the input (of the transpose operation). I assume the former. So we should be clearer.

dwf replied:
Or maybe not, given the first line of apply(). Hmm...

dwf replied:
err _allocate

dwf commented on blocks/bricks/conv.py:
```diff
@@ -177,6 +180,144 @@ def num_output_channels(self):
+        in possibly faster execution times.
+    step : tuple, optional
+        The step (or stride) with which to slide the filters over the
+        image. Defaults to (1, 1).
```
This has a particular strange meaning in the case of this brick. Again, it's the stride of the original un-transposed convolution.

dwf commented on blocks/bricks/conv.py:
```diff
@@ -396,6 +537,54 @@ def _push_allocation_config(self):
+    def _push_allocation_config(self):
+        super(ConvolutionalTransposeActivation, self)._push_allocation_config()
+        self.convolution.step = self.step
+        self.convolution.original_image_size = self.original_image_size
```
Seems  like more stuff should get pushed?

vdumoulin replied:
It's handled by the `_AllocationMixin` superclass.

dwf replied:
Ok good.

vdumoulin commented:
@dwf 0a5f809 addresses your comments regarding the clarity of parameter description.

dwf commented on blocks/bricks/conv.py:
```diff
@@ -177,6 +180,59 @@ def num_output_channels(self):
+
+    Parameters
+    ----------
+    image_size : tuple, optional
```
I think this should go last if anywhere. Probably also should note that this is the input to the *transpose*, output of the original convolution.

dwf commented:
`ConvolutionalTransposeActivation` needs a test.

Other than that and my one comment about the `image_size` documentation, LGTM.

dwf commented on blocks/bricks/conv.py:
```diff
@@ -1,5 +1,8 @@
-from theano.tensor.nnet.conv import conv2d, get_conv_output_shape
-from theano.tensor.signal.downsample import max_pool_2d, DownsampleFactorMax
+from theano import tensor
+from theano.tensor.nnet import conv2d
```
Oh, is this actually using abstract conv2d or just regular old conv2d?

vdumoulin replied:
No, the old `conv2d` is located in `theano.tensor.nnet.conv`. This one aliases to `theano.tensor.nnet.abstract_conv.conv2d`.

dwf replied:
Got it.

vdumoulin commented:
@dwf 87d143f addresses your constructor arg reordering suggestion and removes more duplicate code, as well as clarifies and moves down the `image_size` description. 648646d adds a test for `ConvolutionalTransposeActivation`.

With that, I think I addressed all your comments. Thank you very much for you feedback, and please let me know if there's anything else I forgot.

dwf commented:
LGTM
